cloudinit/util.py

```diff
@@ -287,15 +287,50 @@ def rand_dict_key(dictionary, postfix=None):
     return newkey
 
 
-def read_conf(fname):
+def read_conf(fname, *, instance_data_file=None) -> Dict:
+    """Read a yaml config with optional template, and convert to dict"""
+    # Avoid circular import
+    from cloudinit.handlers.jinja_template import (
+        JinjaLoadError,
+        NotJinjaError,
+        render_jinja_payload_from_file,
+    )
+
     try:
-        return load_yaml(load_file(fname), default={})
+        config_file = load_file(fname)
     except IOError as e:
         if e.errno == ENOENT:
             return {}
         else:
             raise
 
+    if instance_data_file and os.path.exists(instance_data_file):
+        try:
+            config_file = render_jinja_payload_from_file(
+                config_file,
+                fname,
+                instance_data_file,
+            )
+            LOG.debug(
+                "Applied instance data in '%s' to "
+                "configuration loaded from '%s'",
+                instance_data_file,
+                fname,
+            )
+        except NotJinjaError:
+            # A log isn't appropriate here as we generally expect most
+            # cloud.cfgs to not be templated. The other path is logged
+            pass
+        except JinjaLoadError:
+            LOG.warning(
```

Comment:
Nit: I would include here the exception message, as `render_jinja_payload_from_file` is handling [OSError](https://github.com/canonical/cloud-init/pull/1722/files#diff-fb04a669500887728f7bbca305f11ba9727ad3ca8adab8e02f2cdddf11058984R113) exceptions and otherwise that info would not reach the log.

I realize this fact (insance_data_file not readable) is very unlikely, as generally this execution path is executed as root, but `render_jinja_payload_from_file`'s exception handling could grow over time, and this log msg won't inform users of those different msgs.